File: python/src/datongzi_rules/patterns/recognizer.py
"""Card combinations and play types for Da Tong Zi game."""

from collections import Counter
from dataclasses import dataclass
from enum import IntEnum
import logging

from ..models.card import Card, Rank, Suit

logger = logging.getLogger(__name__)


class PlayType(IntEnum):
    """
    Play types in order of strength.
    Higher values beat lower values, with special rules for some types.
    """

    SINGLE = 1  # 单牌
    PAIR = 2  # 对子
    CONSECUTIVE_PAIRS = 3  # 连对 (2+ pairs in sequence)
    TRIPLE = 4  # 三张
    TRIPLE_WITH_TWO = 5  # 三带二
    AIRPLANE = 6  # 飞机 (consecutive triples)
    AIRPLANE_WITH_WINGS = 7  # 飞机带翅膀
    BOMB = 8  # 炸弹 (4+ same rank)
    TONGZI = 9  # 筒子 (3 same rank same suit)
    DIZHA = 10  # 地炸 (2 of each suit for same rank)


@dataclass(frozen=True)
class PlayPattern:
    """Represents a recognized pattern of cards."""

    play_type: PlayType
    primary_rank: Rank  # Main rank of the pattern
    primary_suit: Suit | None = None  # For suit-dependent patterns
    secondary_ranks: list[Rank] | None = None  # For consecutive patterns
    card_count: int = 0  # Total number of cards
    strength: int = 0  # Calculated strength for comparison

    def __post_init__(self) -> None:
        """Initialize computed fields."""
        if self.secondary_ranks is None:
            object.__setattr__(self, "secondary_ranks", [])


class PatternRecognizer:
    """Recognizes and analyzes card patterns."""

    @staticmethod
    def analyze_cards(cards: list[Card]) -> PlayPattern | None:
        """
        Analyze a list of cards and return the recognized pattern.

        Args:
            cards: List of cards to analyze

        Returns:
            PlayPattern if valid pattern found, None otherwise
        """
        if not cards:
            return None

        # Sort cards for easier analysis
        sorted_cards = sorted(cards)

        # Count cards by rank
        rank_counts = Counter(card.rank for card in cards)

        # Count cards by suit and rank (for special patterns)
        suit_rank_counts = Counter((card.suit, card.rank) for card in cards)

        logger.debug(
            f"Analyzing card pattern: card_count={len(cards)}, "
            f"rank_counts={dict(rank_counts)}, "
            f"cards={[str(c) for c in sorted_cards]}"
        )

        # Check for special patterns first (highest priority)
        pattern = PatternRecognizer._check_dizha(cards, suit_rank_counts, rank_counts)
        if pattern:
            return pattern

        pattern = PatternRecognizer._check_tongzi(cards, suit_rank_counts, rank_counts)
        if pattern:
            return pattern

        pattern = PatternRecognizer._check_bomb(cards, rank_counts)
        if pattern:
            return pattern

        # Check for airplane patterns
        # IMPORTANT: Check pure AIRPLANE first, then AIRPLANE_WITH_WINGS
        # This ensures that if all cards form a valid airplane, we don't
        # misidentify it as airplane with wings
        pattern = PatternRecognizer._check_airplane(cards, rank_counts)
        if pattern:
            return pattern

        pattern = PatternRecognizer._check_airplane_with_wings(cards, rank_counts)
        if pattern:
            return pattern

        # Check for basic patterns
        pattern = PatternRecognizer._check_triple_with_two(cards, rank_counts)
        if pattern:
            return pattern

        pattern = PatternRecognizer._check_triple(cards, rank_counts)
        if pattern:
            return pattern

        pattern = PatternRecognizer._check_consecutive_pairs(cards, rank_counts)
        if pattern:
            return pattern

        pattern = PatternRecognizer._check_pair(cards, rank_counts)
        if pattern:
            return pattern

        pattern = PatternRecognizer._check_single(cards, rank_counts)
        if pattern:
            return pattern

        logger.debug(f"No valid pattern recognized: cards={[str(c) for c in cards]}")
        return None

    @staticmethod
    def _check_single(
        cards: list[Card], rank_counts: Counter[Rank]
    ) -> PlayPattern | None:
        """Check for single card pattern."""
        if len(cards) != 1:
            return None

        card = cards[0]
        return PlayPattern(
            play_type=PlayType.SINGLE,
            primary_rank=card.rank,
            primary_suit=card.suit,
            card_count=1,
            strength=card.rank.value,
        )

    @staticmethod
    def _check_pair(
        cards: list[Card], rank_counts: Counter[Rank]
    ) -> PlayPattern | None:
        """Check for pair pattern."""
        if len(cards) != 2 or len(rank_counts) != 1:
            return None

        rank = list(rank_counts.keys())[0]
        if rank_counts[rank] != 2:
            return None

        return PlayPattern(
            play_type=PlayType.PAIR,
            primary_rank=rank,
            card_count=2,
            strength=rank.value,
        )

    @staticmethod
    def _check_consecutive_pairs(
        cards: list[Card], rank_counts: Counter[Rank]
    ) -> PlayPattern | None:
        """Check for consecutive pairs pattern (连对)."""
        if len(cards) < 4 or len(cards) % 2 != 0:
            return None

        # All ranks must have exactly 2 cards
        if any(count != 2 for count in rank_counts.values()):
            return None

        ranks = sorted(rank_counts.keys(), key=lambda r: r.value)

        # Check if ranks are consecutive
        if not PatternRecognizer._are_consecutive(ranks):
            return None

        return PlayPattern(
            play_type=PlayType.CONSECUTIVE_PAIRS,
            primary_rank=ranks[-1],  # Highest rank
            secondary_ranks=ranks,
            card_count=len(cards),
            strength=ranks[-1].value * 1000 + len(ranks),
        )

    @staticmethod
    def _check_triple(
        cards: list[Card], rank_counts: Counter[Rank]
    ) -> PlayPattern | None:
        """Check for triple pattern."""
        if len(cards) != 3 or len(rank_counts) != 1:
            return None

        rank = list(rank_counts.keys())[0]
        if rank_counts[rank] != 3:
            return None

        return PlayPattern(
            play_type=PlayType.TRIPLE,
            primary_rank=rank,
            card_count=3,
            strength=rank.value,
        )

    @staticmethod
    def _check_triple_with_two(
        cards: list[Card], rank_counts: Counter[Rank]
    ) -> PlayPattern | None:
        """Check for triple with two pattern (三带二)."""
        if len(cards) != 5 or len(rank_counts) != 2:
            return None

        counts = list(rank_counts.values())
        if not (3 in counts and 2 in counts):
            return None

        # Find the triple rank
        triple_rank = None
        for rank, count in rank_counts.items():
            if count == 3:
                triple_rank = rank
                break

        if triple_rank is None:
            return None

        return PlayPattern(
            play_type=PlayType.TRIPLE_WITH_TWO,
            primary_rank=triple_rank,
            card_count=5,
            strength=triple_rank.value,
        )

    @staticmethod
    def _check_airplane(
        cards: list[Card], rank_counts: Counter[Rank]
    ) -> PlayPattern | None:
        """Check for airplane pattern (consecutive triples)."""
        if len(cards) < 6 or len(cards) % 3 != 0:
            return None

        # All ranks must have exactly 3 cards
        if any(count != 3 for count in rank_counts.values()):
            return None

        ranks = sorted(rank_counts.keys(), key=lambda r: r.value)

        # Check if ranks are consecutive
        if not PatternRecognizer._are_consecutive(ranks):
            return None

        return PlayPattern(
            play_type=PlayType.AIRPLANE,
            primary_rank=ranks[-1],  # Highest rank
            secondary_ranks=ranks,
            card_count=len(cards),
            strength=ranks[-1].value * 1000 + len(ranks),
        )

    @staticmethod
    def _check_airplane_with_wings(
        cards: list[Card], rank_counts: Counter[Rank]
    ) -> PlayPattern | None:
        """Check for airplane with wings pattern (飞机带翅膀).

        规则：N组连续三张（飞机主体）+ K张任意牌（翅膀）
        其中 N <= K <= 2N
        翅膀可以是任意牌（单张、对子、三张、炸弹等）

        关键：贪心选择最大的连续飞机组（如果翅膀中有3张同rank能扩展飞机，应扩展）
        """
        if len(cards) < 8:  # Minimum: 2 triples (6) + 2 wings (2)
            return None

        # Find all ranks with at least 3 cards
        triple_candidates = [rank for rank, count in rank_counts.items() if count >= 3]

        if len(triple_candidates) < 2:
            return None

        # Sort candidates by rank value
        sorted_candidates = sorted(triple_candidates, key=lambda r: r.value)

        # Strategy: Greedily select the LARGEST consecutive triple sequence
        # This ensures that if wings contain 3+ cards of same rank that can extend
        # the airplane, we prioritize the extended airplane
        best_airplane = None
        best_wing_count = None

        # Try all possible consecutive triple combinations, preferring larger airplanes
        for length in range(len(sorted_candidates), 1, -1):  # Start from longest
            for i in range(len(sorted_candidates) - length + 1):
                candidate_ranks = sorted_candidates[i:i + length]

                if PatternRecognizer._are_consecutive(candidate_ranks):
                    num_triples = len(candidate_ranks)
                    triple_cards = num_triples * 3
                    wing_cards = len(cards) - triple_cards

                    # Check if wing count is valid: N <= wings <= 2N
                    if num_triples <= wing_cards <= 2 * num_triples:
                        # Found valid combination, return immediately (greedy)
                        return PlayPattern(
                            play_type=PlayType.AIRPLANE_WITH_WINGS,
                            primary_rank=candidate_ranks[-1],
                            secondary_ranks=candidate_ranks,
                            card_count=len(cards),
                            strength=candidate_ranks[-1].value * 1000 + len(candidate_ranks),
                        )

        return None

    @staticmethod
    def _check_bomb(
        cards: list[Card], rank_counts: Counter[Rank]
    ) -> PlayPattern | None:
        """Check for bomb pattern (4+ same rank)."""
        if len(cards) < 4 or len(rank_counts) != 1:
            return None

        rank = list(rank_counts.keys())[0]
        count = rank_counts[rank]

        if count < 4:
            return None

        return PlayPattern(
            play_type=PlayType.BOMB,
            primary_rank=rank,
            card_count=count,
            strength=rank.value * 1000 + count,  # Higher count beats same rank
        )

    @staticmethod
    def _check_tongzi(
        cards: list[Card],
        suit_rank_counts: Counter[tuple[Suit, Rank]],
        rank_counts: Counter[Rank],
    ) -> PlayPattern | None:
        """Check for tongzi pattern (3 same rank same suit)."""
        if len(cards) != 3 or len(rank_counts) != 1:
            return None

        # Must have exactly one suit-rank combination with 3 cards
        if len(suit_rank_counts) != 1 or list(suit_rank_counts.values())[0] != 3:
            return None

        suit, rank = list(suit_rank_counts.keys())[0]

        return PlayPattern(
            play_type=PlayType.TONGZI,
            primary_rank=rank,
            primary_suit=suit,
            card_count=3,
            strength=rank.value * 10000 + suit.value * 1000,  # Suit matters for tongzi
        )

    @staticmethod
    def _check_dizha(
        cards: list[Card],
        suit_rank_counts: Counter[tuple[Suit, Rank]],
        rank_counts: Counter[Rank],
    ) -> PlayPattern | None:
        """Check for dizha pattern (2 of each suit for same rank)."""
        if len(cards) != 8 or len(rank_counts) != 1:
            return None

        rank = list(rank_counts.keys())[0]

        # Must have exactly 2 cards of each suit for this rank
        suits_for_rank = [suit for (suit, r) in suit_rank_counts.keys() if r == rank]
        if len(suits_for_rank) != 4:  # All 4 suits
            return None

        # Each suit must have exactly 2 cards
        for suit in Suit:
            if suit_rank_counts.get((suit, rank), 0) != 2:
                return None

        return PlayPattern(
            play_type=PlayType.DIZHA,
            primary_rank=rank,
            card_count=8,
            strength=rank.value * 100000,  # Dizha is very strong
        )

    @staticmethod
    def _are_consecutive(ranks: list[Rank]) -> bool:
        """Check if ranks are consecutive, handling wrap-around for A-2."""
        if len(ranks) <= 1:
            return True

        # Convert to values for comparison
        values = [rank.value for rank in ranks]

        # Check normal consecutive sequence
        for i in range(1, len(values)):
            if values[i] != values[i - 1] + 1:
                return False

        return True


class PlayValidator:
    """Validates plays according to Da Tong Zi rules."""

    @staticmethod
    def can_beat_play(new_cards: list[Card], current_play: PlayPattern | None) -> bool:
        """
        Check if new cards can beat the current play.

        Args:
            new_cards: Cards being played
            current_play: Current play to beat (None if starting new round)

        Returns:
            True if new cards can beat current play
        """
        if current_play is None:
            # Starting new round - any valid pattern is allowed
            pattern = PatternRecognizer.analyze_cards(new_cards)
            return pattern is not None

        new_pattern = PatternRecognizer.analyze_cards(new_cards)
        if new_pattern is None:
            logger.debug("New play has no valid pattern")
            return False

        return PlayValidator._compare_patterns(new_pattern, current_play)

    @staticmethod
    def _compare_patterns(
        new_pattern: PlayPattern, current_pattern: PlayPattern
    ) -> bool:
        """
        Compare two patterns to see if new pattern beats current pattern.

        Returns:
            True if new_pattern beats current_pattern
        """
        # Special cases: Dizha beats everything, Tongzi beats Bomb
        if new_pattern.play_type == PlayType.DIZHA:
            if current_pattern.play_type != PlayType.DIZHA:
                return True
            # Dizha vs Dizha: compare ranks
            return new_pattern.primary_rank.value > current_pattern.primary_rank.value

        if current_pattern.play_type == PlayType.DIZHA:
            return False  # Nothing beats Dizha except higher Dizha

        if new_pattern.play_type == PlayType.TONGZI:
            if current_pattern.play_type == PlayType.BOMB:
                return True  # Tongzi beats Bomb
            elif current_pattern.play_type != PlayType.TONGZI:
                return False  # Tongzi can only beat Bomb or other Tongzi
            # Tongzi vs Tongzi: compare by rank, then by suit
            if new_pattern.primary_rank.value > current_pattern.primary_rank.value:
                return True
            elif new_pattern.primary_rank.value == current_pattern.primary_rank.value:
                # Both suits must not be None for comparison
                if (
                    new_pattern.primary_suit is not None
                    and current_pattern.primary_suit is not None
                ):
                    return (
                        new_pattern.primary_suit.value
                        > current_pattern.primary_suit.value
                    )
                return False
            return False

        if current_pattern.play_type == PlayType.TONGZI:
            return False  # Only Tongzi or Dizha can beat Tongzi

        if new_pattern.play_type == PlayType.BOMB:
            if current_pattern.play_type != PlayType.BOMB:
                return True  # Bomb beats non-bomb
            # Bomb vs Bomb: compare by rank first, then count
            if new_pattern.primary_rank.value > current_pattern.primary_rank.value:
                return True
            elif new_pattern.primary_rank.value == current_pattern.primary_rank.value:
                return new_pattern.card_count > current_pattern.card_count
            return False

        if current_pattern.play_type == PlayType.BOMB:
            return False  # Only Bomb/Tongzi/Dizha can beat Bomb

        # Same type comparison
        if new_pattern.play_type != current_pattern.play_type:
            return False

        # For same types, compare by card count first (for consecutive patterns)
        if new_pattern.play_type in [
            PlayType.CONSECUTIVE_PAIRS,
            PlayType.AIRPLANE,
            PlayType.AIRPLANE_WITH_WINGS,
        ]:
            new_ranks = new_pattern.secondary_ranks or []
            current_ranks = current_pattern.secondary_ranks or []
            if len(new_ranks) != len(current_ranks):
                return False

        # Compare by strength
        return new_pattern.strength > current_pattern.strength

File: rust/datongzi-rules/src/patterns/recognizer.rs
//! Pattern recognition logic for card combinations.

use std::collections::HashMap;

use crate::models::{Card, Rank, Suit};
use super::{PlayPattern, PlayType};

/// Recognizes and analyzes card patterns.
pub struct PatternRecognizer;

impl PatternRecognizer {
    /// Analyze a list of cards and return the recognized pattern.
    ///
    /// Returns `None` if no valid pattern is found.
    ///
    /// # Arguments
    ///
    /// * `cards` - Slice of cards to analyze
    ///
    /// # Returns
    ///
    /// `Some(PlayPattern)` if a valid pattern is recognized, `None` otherwise.
    #[must_use]
    pub fn analyze_cards(cards: &[Card]) -> Option<PlayPattern> {
        if cards.is_empty() {
            return None;
        }

        // Sort cards for easier analysis
        let mut sorted_cards = cards.to_vec();
        sorted_cards.sort();

        // Count cards by rank
        let mut rank_counts: HashMap<Rank, usize> = HashMap::new();
        for card in cards {
            *rank_counts.entry(card.rank).or_insert(0) += 1;
        }

        // Count cards by (suit, rank) for special patterns
        let mut suit_rank_counts: HashMap<(Suit, Rank), usize> = HashMap::new();
        for card in cards {
            *suit_rank_counts.entry((card.suit, card.rank)).or_insert(0) += 1;
        }

        // Check for special patterns first (highest priority)
        if let Some(pattern) = Self::check_dizha(cards, &suit_rank_counts, &rank_counts) {
            return Some(pattern);
        }

        if let Some(pattern) = Self::check_tongzi(cards, &suit_rank_counts, &rank_counts) {
            return Some(pattern);
        }

        if let Some(pattern) = Self::check_bomb(cards, &rank_counts) {
            return Some(pattern);
        }

        // Check for airplane patterns
        // IMPORTANT: Check pure AIRPLANE first, then AIRPLANE_WITH_WINGS
        if let Some(pattern) = Self::check_airplane(cards, &rank_counts) {
            return Some(pattern);
        }

        if let Some(pattern) = Self::check_airplane_with_wings(cards, &rank_counts) {
            return Some(pattern);
        }

        // Check for basic patterns
        if let Some(pattern) = Self::check_triple_with_two(cards, &rank_counts) {
            return Some(pattern);
        }

        if let Some(pattern) = Self::check_triple(cards, &rank_counts) {
            return Some(pattern);
        }

        if let Some(pattern) = Self::check_consecutive_pairs(cards, &rank_counts) {
            return Some(pattern);
        }

        if let Some(pattern) = Self::check_pair(cards, &rank_counts) {
            return Some(pattern);
        }

        if let Some(pattern) = Self::check_single(cards, &rank_counts) {
            return Some(pattern);
        }

        None
    }

    /// Check for single card pattern.
    fn check_single(cards: &[Card], _rank_counts: &HashMap<Rank, usize>) -> Option<PlayPattern> {
        if cards.len() != 1 {
            return None;
        }

        let card = cards[0];
        Some(PlayPattern::new(
            PlayType::Single,
            card.rank,
            Some(card.suit),
            vec![],
            1,
            u32::from(card.rank.value()),
        ))
    }

    /// Check for pair pattern.
    fn check_pair(cards: &[Card], rank_counts: &HashMap<Rank, usize>) -> Option<PlayPattern> {
        if cards.len() != 2 || rank_counts.len() != 1 {
            return None;
        }

        let (&rank, &count) = rank_counts.iter().next()?;
        if count != 2 {
            return None;
        }

        Some(PlayPattern::new(
            PlayType::Pair,
            rank,
            None,
            vec![],
            2,
            u32::from(rank.value()),
        ))
    }

    /// Check for consecutive pairs pattern (连对).
    fn check_consecutive_pairs(
        cards: &[Card],
        rank_counts: &HashMap<Rank, usize>,
    ) -> Option<PlayPattern> {
        if cards.len() < 4 || cards.len() % 2 != 0 {
            return None;
        }

        // All ranks must have exactly 2 cards
        if rank_counts.values().any(|&count| count != 2) {
            return None;
        }

        let mut ranks: Vec<Rank> = rank_counts.keys().copied().collect();
        ranks.sort_by_key(|r| r.value());

        // Check if ranks are consecutive
        if !Self::are_consecutive(&ranks) {
            return None;
        }

        let highest_rank = *ranks.last()?;
        let ranks_len = ranks.len();
        Some(PlayPattern::new(
            PlayType::ConsecutivePairs,
            highest_rank,
            None,
            ranks,
            cards.len(),
            u32::from(highest_rank.value()) * 1000 + ranks_len as u32,
        ))
    }

    /// Check for triple pattern.
    fn check_triple(cards: &[Card], rank_counts: &HashMap<Rank, usize>) -> Option<PlayPattern> {
        if cards.len() != 3 || rank_counts.len() != 1 {
            return None;
        }

        let (&rank, &count) = rank_counts.iter().next()?;
        if count != 3 {
            return None;
        }

        Some(PlayPattern::new(
            PlayType::Triple,
            rank,
            None,
            vec![],
            3,
            u32::from(rank.value()),
        ))
    }

    /// Check for triple with two pattern (三带二).
    fn check_triple_with_two(
        cards: &[Card],
        rank_counts: &HashMap<Rank, usize>,
    ) -> Option<PlayPattern> {
        if cards.len() != 5 || rank_counts.len() != 2 {
            return None;
        }

        let counts: Vec<usize> = rank_counts.values().copied().collect();
        if !(counts.contains(&3) && counts.contains(&2)) {
            return None;
        }

        // Find the triple rank
        let triple_rank = rank_counts
            .iter()
            .find_map(|(&rank, &count)| if count == 3 { Some(rank) } else { None })?;

        Some(PlayPattern::new(
            PlayType::TripleWithTwo,
            triple_rank,
            None,
            vec![],
            5,
            u32::from(triple_rank.value()),
        ))
    }

    /// Check for airplane pattern (consecutive triples).
    fn check_airplane(cards: &[Card], rank_counts: &HashMap<Rank, usize>) -> Option<PlayPattern> {
        if cards.len() < 6 || cards.len() % 3 != 0 {
            return None;
        }

        // All ranks must have exactly 3 cards
        if rank_counts.values().any(|&count| count != 3) {
            return None;
        }

        let mut ranks: Vec<Rank> = rank_counts.keys().copied().collect();
        ranks.sort_by_key(|r| r.value());

        // Check if ranks are consecutive
        if !Self::are_consecutive(&ranks) {
            return None;
        }

        let highest_rank = *ranks.last()?;
        let ranks_len = ranks.len();
        Some(PlayPattern::new(
            PlayType::Airplane,
            highest_rank,
            None,
            ranks,
            cards.len(),
            u32::from(highest_rank.value()) * 1000 + ranks_len as u32,
        ))
    }

    /// Check for airplane with wings pattern (飞机带翅膀).
    ///
    /// Rules: N consecutive triples + K wing cards
    /// where N <= K <= 2N
    /// Wings can be any cards (singles, pairs, triples, bombs, etc.)
    ///
    /// Key: Greedily select the LARGEST consecutive triple sequence
    fn check_airplane_with_wings(
        cards: &[Card],
        rank_counts: &HashMap<Rank, usize>,
    ) -> Option<PlayPattern> {
        if cards.len() < 8 {
            // Minimum: 2 triples (6) + 2 wings (2)
            return None;
        }

        // Find all ranks with at least 3 cards
        let mut triple_candidates: Vec<Rank> = rank_counts
            .iter()
            .filter(|(_, &count)| count >= 3)
            .map(|(&rank, _)| rank)
            .collect();

        if triple_candidates.len() < 2 {
            return None;
        }

        // Sort candidates by rank value
        triple_candidates.sort_by_key(|r| r.value());

        // Strategy: Greedily select the LARGEST consecutive triple sequence
        // Try all possible consecutive triple combinations, preferring larger airplanes
        for length in (2..=triple_candidates.len()).rev() {
            // Start from longest
            for i in 0..=triple_candidates.len() - length {
                let candidate_ranks = &triple_candidates[i..i + length];

                if Self::are_consecutive(candidate_ranks) {
                    let num_triples = candidate_ranks.len();
                    let triple_cards = num_triples * 3;
                    let wing_cards = cards.len() - triple_cards;

                    // Check if wing count is valid: N <= wings <= 2N
                    if wing_cards >= num_triples && wing_cards <= 2 * num_triples {
                        let highest_rank = *candidate_ranks.last()?;
                        return Some(PlayPattern::new(
                            PlayType::AirplaneWithWings,
                            highest_rank,
                            None,
                            candidate_ranks.to_vec(),
                            cards.len(),
                            u32::from(highest_rank.value()) * 1000 + candidate_ranks.len() as u32,
                        ));
                    }
                }
            }
        }

        None
    }

    /// Check for bomb pattern (4+ same rank).
    fn check_bomb(cards: &[Card], rank_counts: &HashMap<Rank, usize>) -> Option<PlayPattern> {
        if cards.len() < 4 || rank_counts.len() != 1 {
            return None;
        }

        let (&rank, &count) = rank_counts.iter().next()?;

        if count < 4 {
            return None;
        }

        Some(PlayPattern::new(
            PlayType::Bomb,
            rank,
            None,
            vec![],
            count,
            u32::from(rank.value()) * 1000 + count as u32,
        ))
    }

    /// Check for tongzi pattern (3 same rank same suit).
    fn check_tongzi(
        cards: &[Card],
        suit_rank_counts: &HashMap<(Suit, Rank), usize>,
        rank_counts: &HashMap<Rank, usize>,
    ) -> Option<PlayPattern> {
        if cards.len() != 3 || rank_counts.len() != 1 {
            return None;
        }

        // Must have exactly one suit-rank combination with 3 cards
        if suit_rank_counts.len() != 1 {
            return None;
        }

        let (&(suit, rank), &count) = suit_rank_counts.iter().next()?;
        if count != 3 {
            return None;
        }

        Some(PlayPattern::new(
            PlayType::Tongzi,
            rank,
            Some(suit),
            vec![],
            3,
            u32::from(rank.value()) * 10000 + u32::from(suit.value()) * 1000,
        ))
    }

    /// Check for dizha pattern (2 of each suit for same rank).
    fn check_dizha(
        cards: &[Card],
        suit_rank_counts: &HashMap<(Suit, Rank), usize>,
        rank_counts: &HashMap<Rank, usize>,
    ) -> Option<PlayPattern> {
        if cards.len() != 8 || rank_counts.len() != 1 {
            return None;
        }

        let &rank = rank_counts.keys().next()?;

        // Must have exactly 2 cards of each suit for this rank
        let suits_for_rank: Vec<Suit> = suit_rank_counts
            .keys()
            .filter(|(_, r)| *r == rank)
            .map(|(s, _)| *s)
            .collect();

        if suits_for_rank.len() != 4 {
            // All 4 suits
            return None;
        }

        // Each suit must have exactly 2 cards
        for suit in [Suit::Diamonds, Suit::Clubs, Suit::Hearts, Suit::Spades] {
            if suit_rank_counts.get(&(suit, rank)) != Some(&2) {
                return None;
            }
        }

        Some(PlayPattern::new(
            PlayType::Dizha,
            rank,
            None,
            vec![],
            8,
            u32::from(rank.value()) * 100000,
        ))
    }

    /// Check if ranks are consecutive.
    fn are_consecutive(ranks: &[Rank]) -> bool {
        if ranks.len() <= 1 {
            return true;
        }

        // Convert to values for comparison
        let values: Vec<u8> = ranks.iter().map(|r| r.value()).collect();

        // Check normal consecutive sequence
        for i in 1..values.len() {
            if values[i] != values[i - 1] + 1 {
                return false;
            }
        }

        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_single_pattern() {
        let cards = vec![Card::new(Suit::Spades, Rank::Ace)];
        let pattern = PatternRecognizer::analyze_cards(&cards);
        assert!(pattern.is_some());
        let pattern = pattern.unwrap();
        assert_eq!(pattern.play_type, PlayType::Single);
        assert_eq!(pattern.primary_rank, Rank::Ace);
        assert_eq!(pattern.strength, 14);
    }

    #[test]
    fn test_pair_pattern() {
        let cards = vec![
            Card::new(Suit::Spades, Rank::Ace),
            Card::new(Suit::Hearts, Rank::Ace),
        ];
        let pattern = PatternRecognizer::analyze_cards(&cards);
        assert!(pattern.is_some());
        let pattern = pattern.unwrap();
        assert_eq!(pattern.play_type, PlayType::Pair);
        assert_eq!(pattern.primary_rank, Rank::Ace);
    }

    #[test]
    fn test_consecutive_pairs() {
        let cards = vec![
            Card::new(Suit::Spades, Rank::Three),
            Card::new(Suit::Hearts, Rank::Three),
            Card::new(Suit::Spades, Rank::Four),
            Card::new(Suit::Hearts, Rank::Four),
        ];
        let pattern = PatternRecognizer::analyze_cards(&cards);
        assert!(pattern.is_some());
        let pattern = pattern.unwrap();
        assert_eq!(pattern.play_type, PlayType::ConsecutivePairs);
        assert_eq!(pattern.primary_rank, Rank::Four);
    }

    #[test]
    fn test_triple_pattern() {
        let cards = vec![
            Card::new(Suit::Spades, Rank::King),
            Card::new(Suit::Hearts, Rank::King),
            Card::new(Suit::Clubs, Rank::King),
        ];
        let pattern = PatternRecognizer::analyze_cards(&cards);
        assert!(pattern.is_some());
        let pattern = pattern.unwrap();
        // Should be Triple, not Tongzi (different suits)
        assert_eq!(pattern.play_type, PlayType::Triple);
        assert_eq!(pattern.primary_rank, Rank::King);
    }

    #[test]
    fn test_tongzi_pattern() {
        let cards = vec![
            Card::new(Suit::Spades, Rank::King),
            Card::new(Suit::Spades, Rank::King),
            Card::new(Suit::Spades, Rank::King),
        ];
        let pattern = PatternRecognizer::analyze_cards(&cards);
        assert!(pattern.is_some());
        let pattern = pattern.unwrap();
        assert_eq!(pattern.play_type, PlayType::Tongzi);
        assert_eq!(pattern.primary_rank, Rank::King);
        assert_eq!(pattern.primary_suit, Some(Suit::Spades));
    }

    #[test]
    fn test_bomb_pattern() {
        let cards = vec![
            Card::new(Suit::Spades, Rank::Ace),
            Card::new(Suit::Hearts, Rank::Ace),
            Card::new(Suit::Clubs, Rank::Ace),
            Card::new(Suit::Diamonds, Rank::Ace),
        ];
        let pattern = PatternRecognizer::analyze_cards(&cards);
        assert!(pattern.is_some());
        let pattern = pattern.unwrap();
        assert_eq!(pattern.play_type, PlayType::Bomb);
        assert_eq!(pattern.primary_rank, Rank::Ace);
        assert_eq!(pattern.card_count, 4);
    }

    #[test]
    fn test_dizha_pattern() {
        let cards = vec![
            Card::new(Suit::Spades, Rank::Two),
            Card::new(Suit::Spades, Rank::Two),
            Card::new(Suit::Hearts, Rank::Two),
            Card::new(Suit::Hearts, Rank::Two),
            Card::new(Suit::Clubs, Rank::Two),
            Card::new(Suit::Clubs, Rank::Two),
            Card::new(Suit::Diamonds, Rank::Two),
            Card::new(Suit::Diamonds, Rank::Two),
        ];
        let pattern = PatternRecognizer::analyze_cards(&cards);
        assert!(pattern.is_some());
        let pattern = pattern.unwrap();
        assert_eq!(pattern.play_type, PlayType::Dizha);
        assert_eq!(pattern.primary_rank, Rank::Two);
    }

    #[test]
    fn test_airplane_pattern() {
        let cards = vec![
            Card::new(Suit::Spades, Rank::Three),
            Card::new(Suit::Hearts, Rank::Three),
            Card::new(Suit::Clubs, Rank::Three),
            Card::new(Suit::Spades, Rank::Four),
            Card::new(Suit::Hearts, Rank::Four),
            Card::new(Suit::Clubs, Rank::Four),
        ];
        let pattern = PatternRecognizer::analyze_cards(&cards);
        assert!(pattern.is_some());
        let pattern = pattern.unwrap();
        assert_eq!(pattern.play_type, PlayType::Airplane);
        assert_eq!(pattern.primary_rank, Rank::Four);
    }

    #[test]
    fn test_are_consecutive() {
        let ranks = vec![Rank::Three, Rank::Four, Rank::Five];
        assert!(PatternRecognizer::are_consecutive(&ranks));

        let ranks = vec![Rank::Three, Rank::Five];
        assert!(!PatternRecognizer::are_consecutive(&ranks));

        let ranks = vec![Rank::Ace];
        assert!(PatternRecognizer::are_consecutive(&ranks));
    }
}
